vendor/bat-native-ads/src/bat/ads/internal/ad_preferences.cc

```diff
@@ -0,0 +1,143 @@
+/* Copyright (c) 2019 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "bat/ads/internal/ad_preferences.h"
+
+#include "bat/ads/internal/filtered_ad.h"
+#include "bat/ads/internal/filtered_category.h"
+#include "bat/ads/internal/flagged_ad.h"
+#include "bat/ads/internal/json_helper.h"
+#include "bat/ads/internal/saved_ad.h"
+
+namespace ads {
+
+AdPreferences::AdPreferences() :
+    filtered_ads({}),
+    filtered_categories({}),
+    saved_ads({}),
+    flagged_ads({}) {}
+
+AdPreferences::AdPreferences(const AdPreferences& prefs) :
+    filtered_ads(prefs.filtered_ads),
+    filtered_categories(prefs.filtered_categories),
+    saved_ads(prefs.saved_ads),
+    flagged_ads(prefs.flagged_ads) {}
+
+AdPreferences::~AdPreferences() = default;
+
+const std::string AdPreferences::ToJson() const {
+  std::string json;
+  SaveToJson(*this, &json);
+  return json;
+}
+
+Result AdPreferences::FromJson(
+    const std::string& json,
+    std::string* error_description) {
+  rapidjson::Document document;
+  document.Parse(json.c_str());
+
+  if (document.HasParseError()) {
+    if (error_description != nullptr) {
+      *error_description = helper::JSON::GetLastError(&document);
+    }
+
+    return FAILED;
+  }
+
+  for (const auto& ad : document["filtered_ads"].GetArray()) {
```

Comment:
@tmancey filtered_ads are ads that the user has indicated they don't like, so we block them based on creative_set_id.